src/transformers/commands/serving.py

```diff
@@ -98,131 +135,193 @@ def register_subcommand(parser: ArgumentParser):
 
-        if not _serve_dependencies_installed:
-            raise RuntimeError(
-                "Using serve command requires FastAPI and uvicorn. "
-                'Please install transformers with [serving]: pip install "transformers[serving]". '
-                "Or install FastAPI and uvicorn separately."
+    def generate(self, app):
+        @app.post("/v1/chat/completions")
+        def _serve(req: ChatCompletionInput):
+            if not req.stream:
+                return {"error": "Only streaming mode is supported."}
+
+            text = self.tokenizer.apply_chat_template(
+                req.messages, tools=req.tools, add_generation_prompt=True, tokenize=False
             )
-        else:
-            logger.info(f"Serving model over {host}:{port}")
-            self._app = FastAPI(
-                routes=[
-                    APIRoute(
-                        "/",
-                        self.model_info,
-                        response_model=ServeModelInfoResult,
-                        response_class=JSONResponse,
-                        methods=["GET"],
-                    ),
-                    APIRoute(
-                        "/tokenize",
-                        self.tokenize,
-                        response_model=ServeTokenizeResult,
-                        response_class=JSONResponse,
-                        methods=["POST"],
-                    ),
-                    APIRoute(
-                        "/detokenize",
-                        self.detokenize,
-                        response_model=ServeDeTokenizeResult,
-                        response_class=JSONResponse,
-                        methods=["POST"],
-                    ),
-                    APIRoute(
-                        "/forward",
-                        self.forward,
-                        response_model=ServeForwardResult,
-                        response_class=JSONResponse,
-                        methods=["POST"],
-                    ),
-                ],
-                timeout=600,
+
+            inputs = self.tokenizer(text, return_tensors="pt").to(self.model.device)["input_ids"]
+            request_id = req.request_id if req.request_id is not None else "req_0"
+
+            generation_streamer = TextIteratorStreamer(self.tokenizer, skip_special_tokens=True, skip_prompt=True)
+
+            if req.generation_config is None:
+                req.generation_config = {}
+
+            generation_config = GenerationConfig(**req.generation_config)
+
+            generation_kwargs = {
+                "inputs": inputs,
+                "attention_mask": torch.ones_like(inputs),
+                "streamer": generation_streamer,
+                "generation_config": generation_config,
+            }
+
+            def stream_response(streamer, _request_id):
+                thread = Thread(target=self.model.generate, kwargs=generation_kwargs)
```

Comment:
Do I understand it correctly that `self.model` is instantiated once and then we use it with a new "TextIteratorStreamer" each time? Out of curiosity, what's the behavior if 2 concurrent requests are sent? Is the second one waiting until the first is entirely completed?